In [3]:
import sys
sys.path.append('../')

from utils.utils import excute_cmd, execute_cmd_realtime
import matplotlib.pyplot as plt

%matplotlib inline

# To allow auto reload to this notebook after modifying any external file imported
%load_ext autoreload
%autoreload 2

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


Training a single using the train masks, lower lr, and more epochs using `ClassifierSegExperiment`. 

In [4]:
command = 'python ../train.py --train_path "../datasets/challenge2/train" \
                        --train_masks_path "../datasets_masks/challenge2/train" \
                        --valid_path "../datasets/challenge2/val" \
                        --experiment_name "ClassifierSegExperiment" \
                        --network_name "VGG16_BN_Attention" \
                        --max_epochs "50" \
                        --base_lr "0.00001" \
                        --batch_size "32" \
                        --verbose "2" \
                        --focal_loss \
                        --multi'

# run the function to excute the command
_ = execute_cmd_realtime(command, log=True)

Excuting training pipeline with 50 epochs...
Loading data with 3 class labels...
Loading the data from ../datasets/challenge2/train
Loading the data from ../datasets/challenge2/val
Using segmentation masks for training...
train_images: 5082, train_masks: 5082, train_labels: 5082
Class weights: [0.62440103 0.84997491 4.50531915]
Creating a classifier experiment using VGG16_BN_Attention network.
Using VGG16_BN_Attention with configurations: num_classes='3', normalize_attn='False'
Using focal loss.
Experiment started.
Epoch: 1 Train batch 10 loss: 0.4747427701950073, 6.3% complete
Epoch: 1 Train batch 20 loss: 0.44776538014411926, 12.6% complete
Epoch: 1 Train batch 30 loss: 0.43278053402900696, 18.9% complete
Epoch: 1 Train batch 40 loss: 0.4322032928466797, 25.2% complete
Epoch: 1 Train batch 50 loss: 0.42653924226760864, 31.4% complete
Epoch: 1 Train batch 60 loss: 0.4082804024219513, 37.7% complete
Epoch: 1 Train batch 70 loss: 0.3972669243812561, 44.0% complete
Epoch: 1 Train batch 8

In [ ]:
# 2023-12-29_0940_VGG16_BN_Attention
command = 'python ../train_cv.py \
                    --train_path "../datasets/challenge2/train" \
                    --train_masks_path "../datasets_masks/challenge2/train" \
                    --valid_path "../datasets/challenge2/val" \
                    --valid_masks_path "../datasets_masks/challenge2/val"\
                    --experiment_name "ClassifierSegExperimentCV" \
                    --network_name "VGG16_BN_Attention" \
                    --max_epochs "50" \
                    --base_lr "0.00001" \
                    --num_folds "5" \
                    --batch_size "32" \
                    --verbose "1" \
                    --focal_loss \
                    --multi'

# run the function to excute the command
_ = execute_cmd_realtime(command, log=True)

Excuting training pipeline with 50 epochs and 5 folds.
Loading data with 3 class labels...
Loading the data from ../datasets/challenge2/train
Loading the data from ../datasets/challenge2/val
Using segmentation masks for training...
train_images: 5082, train_masks: 5082, train_labels: 5082
val_images: 1270, val_masks: 1270, val_labels: 1270
Class weights: [0.62427817 0.85023427 4.50443262]
Creating a classifier experiment using VGG16_BN_Attention network.
Using VGG16_BN_Attention with configurations: num_classes='3', normalize_attn='False'
Using focal loss.
Using dice loss.


In [3]:
# BEST RESULT CHALLENGE 2
command = 'python ../inference.py --test_path "../datasets/challenge2/val" \
                    --experiment_name "ClassifierSegExperimentCV" \
                    --network_name "VGG16_BN_Attention" \
                    --max_epochs "50" \
                    --base_lr "0.00001" \
                    --batch_size "32" \
                    --timeframe "2023-12-29_0940" \
                    --verbose "2" \
                    --multi \
                    --ensemble \
                    --report'

# run the function to excute the command
_ = execute_cmd_realtime(command, log=True)

Excuting inference script...
Namespace(test_path='../datasets/challenge2/val', output='outputs', experiment_name='ClassifierSegExperimentCV', network_name='VGG16_BN_Attention', max_epochs=50, batch_size=32, base_lr=1e-05, img_size=224, seed=42, timeframe='2023-12-29_0940', verbose=2, multi=True, report=True, ensemble=True)
Constructed output path: outputs/ClassifierSegExperimentCV_224_epo50_bs32_lr1e-05_s42/2023-12-29_0940_VGG16_BN_Attention. Searching for models...
Found 5 models. Starting loading the models.
Loading data with 3 class labels from ../datasets/challenge2/val path...
Dataset labels: {'mel': 0, 'bcc': 1, 'scc': 2} dictionary.
Loading the data from ../datasets/challenge2/val
Dataset length: 1270
Using VGG16_BN_Attention with configurations: num_classes='3', normalize_attn='False'
Using VGG16_BN_Attention with configurations: num_classes='3', normalize_attn='False'
Using VGG16_BN_Attention with configurations: num_classes='3', normalize_attn='False'
Using VGG16_BN_Attention

In [2]:
# Running inference on valid split for best results challenge 2 (ensemble) to see if everything is ok
command = 'python ../inference.py --test_path "../datasets/challenge2/val" \
                    --experiment_name "ClassifierSegExperimentCV" \
                    --network_name "VGG16_BN_Attention" \
                    --max_epochs "50" \
                    --base_lr "0.00001" \
                    --batch_size "32" \
                    --timeframe "2023-12-29_0940" \
                    --verbose "2" \
                    --multi \
                    --ensemble \
                    --report'

# run the function to excute the command
_ = execute_cmd_realtime(command, log=True)

Excuting inference script...
Namespace(test_path='../datasets/challenge2/val', output='outputs', experiment_name='ClassifierSegExperimentCV', network_name='VGG16_BN_Attention', max_epochs=50, batch_size=32, base_lr=1e-05, img_size=224, seed=42, timeframe='2023-12-29_0940', verbose=2, multi=True, report=True, ensemble=True, combination_strategy='majority_vote', upscale_factor=8, gradcam=False)
Constructed output path: outputs/ClassifierSegExperimentCV_224_epo50_bs32_lr1e-05_s42/2023-12-29_0940_VGG16_BN_Attention. Searching for models...
Found 5 models.
Loading data with 3 class labels from ../datasets/challenge2/val path...
Dataset labels: {'mel': 0, 'bcc': 1, 'scc': 2} dictionary.
Loading the data from ../datasets/challenge2/val
Dataset length: 1270
Using VGG16_BN_Attention with configurations: num_classes='3', normalize_attn='False'
Using VGG16_BN_Attention with configurations: num_classes='3', normalize_attn='False'
Using VGG16_BN_Attention with configurations: num_classes='3', norma

In [3]:
# Running inference on test split for best results challenge 2 (ensemble)
# Majority Voting Ensemble
# Note that for ensuring that shuffle is False, we also exported the prediction filenames to the same csv file
# that filenames column has to be deleted manually before submission
command = 'python ../inference.py --test_path "../datasets/challenge2/test" \
                    --experiment_name "ClassifierSegExperimentCV" \
                    --network_name "VGG16_BN_Attention" \
                    --max_epochs "50" \
                    --base_lr "0.00001" \
                    --batch_size "32" \
                    --timeframe "2023-12-29_0940" \
                    --verbose "2" \
                    --multi \
                    --ensemble'

# run the function to excute the command
_ = execute_cmd_realtime(command, log=True)

Excuting inference script...
Namespace(test_path='../datasets/challenge2/test', output='outputs', experiment_name='ClassifierSegExperimentCV', network_name='VGG16_BN_Attention', max_epochs=50, batch_size=32, base_lr=1e-05, img_size=224, seed=42, timeframe='2023-12-29_0940', verbose=2, multi=True, report=False, ensemble=True, combination_strategy='majority_vote', upscale_factor=8, gradcam=False)
Constructed output path: outputs/ClassifierSegExperimentCV_224_epo50_bs32_lr1e-05_s42/2023-12-29_0940_VGG16_BN_Attention. Searching for models...
Found 5 models.
Loading data with unknown class labels from ../datasets/challenge2/test path...
Dataset labels: {'testX': -911, 'testY': -922, 'testZ': -933} dictionary.
Loading the data from ../datasets/challenge2/test
Forcing n_classes to be based on the length of the labels dictionary...
Before: 1
After: 3
Dataset length: 2121
Using VGG16_BN_Attention with configurations: num_classes='3', normalize_attn='False'
Using VGG16_BN_Attention with configur

In [5]:
# Inferencing only the best 3 folds models from the baseline in the same folder << on the test set >>
# BEST 3 FOLDS CHALLENGE 2 <<SUBMISSION>>
command = 'python ../inference.py --test_path "../datasets/challenge2/test" \
                    --experiment_name "BestK3" \
                    --network_name "VGG16_BN_Attention" \
                    --max_epochs "50" \
                    --base_lr "0.00001" \
                    --batch_size "32" \
                    --timeframe "2023-12-29_0940" \
                    --verbose "2" \
                    --multi \
                    --ensemble'

# run the function to excute the command
_ = execute_cmd_realtime(command, log=True)

Excuting inference script...
Namespace(test_path='../datasets/challenge2/test', output='outputs', experiment_name='BestK3', network_name='VGG16_BN_Attention', max_epochs=50, batch_size=32, base_lr=1e-05, img_size=224, seed=42, timeframe='2023-12-29_0940', verbose=2, multi=True, report=False, ensemble=True, combination_strategy='majority_vote', upscale_factor=8, gradcam=False)
Constructed output path: outputs/BestK3_224_epo50_bs32_lr1e-05_s42/2023-12-29_0940_VGG16_BN_Attention. Searching for models...
Found 3 models.
Loading data with unknown class labels from ../datasets/challenge2/test path...
Dataset labels: {'testX': -911, 'testY': -922, 'testZ': -933} dictionary.
Loading the data from ../datasets/challenge2/test
Forcing n_classes to be based on the length of the labels dictionary...
Before: 1
After: 3
Dataset length: 2121
Using VGG16_BN_Attention with configurations: num_classes='3', normalize_attn='False'
Using VGG16_BN_Attention with configurations: num_classes='3', normalize_att

In [2]:
# BEST RESULT CHALLENGE 2
# command = 'python ../inference.py --test_path "../datasets/challenge2/val" \
#                     --experiment_name "ClassifierSegExperimentCV" \
#                     --network_name "VGG16_BN_Attention" \
#                     --max_epochs "50" \
#                     --base_lr "0.00001" \
#                     --batch_size "32" \
#                     --timeframe "2023-12-29_0940" \
#                     --verbose "2" \
#                     --multi \
#                     --ensemble \
#                     --report'

# command = 'python ../inference.py --test_path "../datasets/challenge2/val" \
#                     --experiment_name "ClassifierSegExperimentCV" \
#                     --network_name "VGG16_BN_Attention" \
#                     --max_epochs "50" \
#                     --base_lr "0.00001" \
#                     --batch_size "32" \
#                     --timeframe "2023-12-29_0940" \
#                     --verbose "2" \
#                     --multi \
#                     --ensemble \
#                     --report \
#                     --combination_strategy "weighted_voting"'

# command = 'python ../inference.py --test_path "../datasets/challenge2/val" \
#                     --experiment_name "ClassifierSegExperimentCV" \
#                     --network_name "VGG16_BN_Attention" \
#                     --max_epochs "50" \
#                     --base_lr "0.00001" \
#                     --batch_size "32" \
#                     --timeframe "2023-12-29_0940" \
#                     --verbose "2" \
#                     --multi \
#                     --ensemble \
#                     --report'

# command = 'python ../inference.py --test_path "../datasets/challenge2/val" \
#                     --experiment_name "ClassifierSegExperimentCV" \
#                     --network_name "VGG16_BN_Attention" \
#                     --max_epochs "50" \
#                     --base_lr "0.00001" \
#                     --batch_size "32" \
#                     --timeframe "2023-12-29_0940" \
#                     --verbose "2" \
#                     --multi \
#                     --ensemble \
#                     --report \
#                     --gradcam'


# run the function to excute the command
_ = execute_cmd_realtime(command, log=True)

Excuting inference script...
Namespace(test_path='../datasets/challenge2/val', output='outputs', experiment_name='ClassifierSegExperimentCV', network_name='VGG16_BN_Attention', max_epochs=50, batch_size=32, base_lr=1e-05, img_size=224, seed=42, timeframe='2023-12-29_0940', verbose=2, multi=True, report=True, ensemble=True, combination_strategy='majority_vote', upscale_factor=8, gradcam=False)
Constructed output path: outputs/ClassifierSegExperimentCV_224_epo50_bs32_lr1e-05_s42/2023-12-29_0940_VGG16_BN_Attention. Searching for models...
Found 5 models.
Loading data with 3 class labels from ../datasets/challenge2/val path...
Dataset labels: {'mel': 0, 'bcc': 1, 'scc': 2} dictionary.
Loading the data from ../datasets/challenge2/val
Dataset length: 1270
Using VGG16_BN_Attention with configurations: num_classes='3', normalize_attn='False'
Using VGG16_BN_Attention with configurations: num_classes='3', normalize_attn='False'
Using VGG16_BN_Attention with configurations: num_classes='3', norma

Changes to hair augmentation.

In [ ]:
# BEST Single Model
# 2024-01-02_2253_VGG16_BN_Attention
command = 'python ../train.py --train_path "../datasets/challenge2/train" \
                        --train_masks_path "../datasets_masks/challenge2/train" \
                        --valid_path "../datasets/challenge2/val" \
                        --experiment_name "ClassifierSegExperiment" \
                        --network_name "VGG16_BN_Attention" \
                        --max_epochs "50" \
                        --base_lr "0.00001" \
                        --batch_size "32" \
                        --verbose "2" \
                        --focal_loss \
                        --multi'

# run the function to excute the command
_ = execute_cmd_realtime(command, log=True)

Excuting training pipeline with 50 epochs...
Loading data with 3 class labels...
Loading the data from ../datasets/challenge2/train
Loading the data from ../datasets/challenge2/val
Using segmentation masks for training...
train_images: 5082, train_masks: 5082, train_labels: 5082
Class weights: [0.62440103 0.84997491 4.50531915]
Creating a classifier experiment using VGG16_BN_Attention network.
Using VGG16_BN_Attention with configurations: num_classes='3', normalize_attn='False'
Using focal loss.
Using dice loss.
Experiment started.
Epoch: 1 Train batch 10 loss: 0.47001636028289795, 6.3% complete
Epoch: 1 Train batch 20 loss: 0.4486575722694397, 12.6% complete


<pre>
2024-01-03 13:00:28.754 | INFO     | experiments.ClassifierSegExperiment:run_test:387 - Test Accuracy: 0.93386
2024-01-03 13:00:28.755 | INFO     | experiments.ClassifierSegExperiment:run_test:388 - Test AUC: 0.98470
2024-01-03 13:00:28.756 | INFO     | experiments.ClassifierSegExperiment:run_test:389 - Test AUC Scores: [0.98784182 0.98899484 0.97724888]
2024-01-03 13:00:28.756 | INFO     | experiments.ClassifierSegExperiment:run_test:390 - Test Kappa: 0.87990
2024-01-03 13:00:28.756 | INFO     | experiments.ClassifierSegExperiment:run_test:394 - Target 0 - Precision: 0.95051, Recall: 0.96313, F-score: 0.95678, Sensitivity: 0.9631268436578171, Specificity: 0.9425675675675675, Support: 678
2024-01-03 13:00:28.757 | INFO     | experiments.ClassifierSegExperiment:run_test:394 - Target 1 - Precision: 0.93267, Recall: 0.94578, F-score: 0.93918, Sensitivity: 0.9457831325301205, Specificity: 0.9559585492227979, Support: 498
2024-01-03 13:00:28.757 | INFO     | experiments.ClassifierSegExperiment:run_test:394 - Target 2 - Precision: 0.79487, Recall: 0.65957, F-score: 0.72093, Sensitivity: 0.6595744680851063, Specificity: 0.9863945578231292, Support: 94
2024-01-03 13:00:28.757 | INFO     | experiments.ClassifierSegExperiment:run_test:396 - 
Classification Report:
            precision    recall  f1-score   support

            0       0.95      0.96      0.96       678
            1       0.93      0.95      0.94       498
            2       0.79      0.66      0.72        94

      accuracy                            0.93      1270
      macro avg       0.89      0.86      0.87      1270
      weighted avg    0.93      0.93      0.93      1270
</pre>

In [ ]:
# 2024-01-03_1221_VGG16_BN_Attention
command = 'python ../train_cv.py \
                    --train_path "../datasets/challenge2/train" \
                    --train_masks_path "../datasets_masks/challenge2/train" \
                    --valid_path "../datasets/challenge2/val" \
                    --valid_masks_path "../datasets_masks/challenge2/val"\
                    --experiment_name "ClassifierSegExperimentCV" \
                    --network_name "VGG16_BN_Attention" \
                    --max_epochs "50" \
                    --base_lr "0.00001" \
                    --num_folds "5" \
                    --batch_size "32" \
                    --verbose "1" \
                    --focal_loss \
                    --multi'

# run the function to excute the command
_ = execute_cmd_realtime(command, log=True)

Excuting training pipeline with 50 epochs and 5 folds.
Loading data with 3 class labels...
Loading the data from ../datasets/challenge2/train
Loading the data from ../datasets/challenge2/val
Using segmentation masks for training...
train_images: 5082, train_masks: 5082, train_labels: 5082
val_images: 1270, val_masks: 1270, val_labels: 1270
Class weights: [0.62427817 0.85023427 4.50443262]
Creating a classifier experiment using VGG16_BN_Attention network.
Using VGG16_BN_Attention with configurations: num_classes='3', normalize_attn='False'
Using focal loss.
Using dice loss.


In [2]:
command = 'python ../inference.py --test_path "../datasets/challenge2/val" \
                    --experiment_name "ClassifierSegExperimentCV" \
                    --network_name "VGG16_BN_Attention" \
                    --max_epochs "50" \
                    --base_lr "0.00001" \
                    --batch_size "32" \
                    --timeframe "2024-01-03_1221" \
                    --verbose "2" \
                    --multi \
                    --ensemble \
                    --report'

# run the function to excute the command
_ = execute_cmd_realtime(command, log=True)

Excuting inference script...
Namespace(test_path='../datasets/challenge2/val', output='outputs', experiment_name='ClassifierSegExperimentCV', network_name='VGG16_BN_Attention', max_epochs=50, batch_size=32, base_lr=1e-05, img_size=224, seed=42, timeframe='2024-01-03_1221', verbose=2, multi=True, report=True, ensemble=True)
Constructed output path: outputs/ClassifierSegExperimentCV_224_epo50_bs32_lr1e-05_s42/2024-01-03_1221_VGG16_BN_Attention. Searching for models...
Found 5 models. Starting loading the models.
Loading data with 3 class labels from ../datasets/challenge2/val path...
Dataset labels: {'mel': 0, 'bcc': 1, 'scc': 2} dictionary.
Loading the data from ../datasets/challenge2/val
Dataset length: 1270
Using VGG16_BN_Attention with configurations: num_classes='3', normalize_attn='False'
Using VGG16_BN_Attention with configurations: num_classes='3', normalize_attn='False'
Using VGG16_BN_Attention with configurations: num_classes='3', normalize_attn='False'
Using VGG16_BN_Attention

In [2]:
command = 'python ../inference.py --test_path "../datasets/challenge2/val" \
                    --experiment_name "ClassifierSegExperimentCV" \
                    --network_name "VGG16_BN_Attention" \
                    --max_epochs "50" \
                    --base_lr "0.00001" \
                    --batch_size "32" \
                    --timeframe "2024-01-03_1221" \
                    --verbose "2" \
                    --multi \
                    --ensemble \
                    --report \
                    --combination_strategy "majority_vote"'

# # run the function to excute the command
_ = execute_cmd_realtime(command, log=True)

Excuting inference script...
Namespace(test_path='../datasets/challenge2/val', output='outputs', experiment_name='ClassifierSegExperimentCV', network_name='VGG16_BN_Attention', max_epochs=50, batch_size=32, base_lr=1e-05, img_size=224, seed=42, timeframe='2024-01-03_1221', verbose=2, multi=True, report=True, ensemble=True, combination_strategy='majority_vote')
Constructed output path: outputs/ClassifierSegExperimentCV_224_epo50_bs32_lr1e-05_s42/2024-01-03_1221_VGG16_BN_Attention. Searching for models...
Found 5 models.
Loading data with 3 class labels from ../datasets/challenge2/val path...
Dataset labels: {'mel': 0, 'bcc': 1, 'scc': 2} dictionary.
Loading the data from ../datasets/challenge2/val
Dataset length: 1270
Using VGG16_BN_Attention with configurations: num_classes='3', normalize_attn='False'
Using VGG16_BN_Attention with configurations: num_classes='3', normalize_attn='False'
Using VGG16_BN_Attention with configurations: num_classes='3', normalize_attn='False'
Using VGG16_BN_

In [7]:
command = 'python ../inference.py --test_path "../datasets/challenge2/val" \
                    --experiment_name "ClassifierSegExperimentCV" \
                    --network_name "VGG16_BN_Attention" \
                    --max_epochs "50" \
                    --base_lr "0.00001" \
                    --batch_size "32" \
                    --timeframe "2024-01-03_1221" \
                    --verbose "2" \
                    --multi \
                    --ensemble \
                    --report \
                    --combination_strategy "weighted_voting"'

# # run the function to excute the command
_ = execute_cmd_realtime(command, log=True)

Excuting inference script...
Namespace(test_path='../datasets/challenge2/val', output='outputs', experiment_name='ClassifierSegExperimentCV', network_name='VGG16_BN_Attention', max_epochs=50, batch_size=32, base_lr=1e-05, img_size=224, seed=42, timeframe='2024-01-03_1221', verbose=2, multi=True, report=True, ensemble=True, combination_strategy='weighted_voting')
Constructed output path: outputs/ClassifierSegExperimentCV_224_epo50_bs32_lr1e-05_s42/2024-01-03_1221_VGG16_BN_Attention. Searching for models...
Found 5 models.
Loading data with 3 class labels from ../datasets/challenge2/val path...
Dataset labels: {'mel': 0, 'bcc': 1, 'scc': 2} dictionary.
Loading the data from ../datasets/challenge2/val
Dataset length: 1270
Combination strategy is weighted_voting. Loading validation data to obtain the weights...
args.val_path: 
Loading data with 3 class labels from ../datasets/challenge2/val path...
Dataset _val_labels: {'mel': 0, 'bcc': 1, 'scc': 2} dictionary.
Loading the data from ../dat